github/checks.go

```diff
@@ -0,0 +1,133 @@
+	PullRequests []*PullRequest  `json:"pull_requests,omitempty"`
+}
+
+// CheckRunOutput represents the output of a CheckRun.
+type CheckRunOutput struct {
+	Title            *string            `json:"title,omitempty"`
+	Summary          *string            `json:"summary,omitempty"`
+	Text             *string            `json:"text,omitempty"`
+	AnnotationsCount *int               `json:"annotations_count,omitempty"`
+	AnnotationsURL   *string            `json:"annotations_url,omitempty"`
+	Annotations      []*CheckAnnotation `json:"annotations,omitempty"`
+	Images           []*CheckImage      `json:"images,omitempty"`
+}
+
+// CheckAnnotation represents an annotation object for a CheckRun output.
+type CheckAnnotation struct {
+	FileName     *string `json:"filename,omitempty"`
+	BlobHRef     *string `json:"blob_href,omitempty"`
+	StartLine    *int    `json:"start_line,omitempty"`
+	EndLine      *int    `json:"end_line,omitempty"`
+	WarningLevel *string `json:"warning_level,omitempty"`
+	Message      *string `json:"message,omitempty"`
+	Title        *string `json:"title,omitempty"`
+	RawDetails   *string `json:"raw_details,omitempty"`
+}
+
+// CheckImage represents an image object for a CheckRun output.
+type CheckImage struct {
+	Alt      *string `json:"alt,omitempty"`
+	ImageURL *string `json:"image_url,omitempty"`
+	Caption  *string `json:"caption,omitempty"`
+}
+
+// CheckSuite represents a suite of check runs.
+type CheckSuite struct {
+	ID *int64 `json:"id,omitempty"`
+}
+
+func (c CheckRun) String() string {
+	return Stringify(c)
+}
+
+// GetCheckRun gets a check-run for a repository.
+//
+// GitHub API docs: https://developer.github.com/v3/checks/runs/#get-a-single-check-run
+func (s *ChecksService) GetCheckRun(ctx context.Context, owner string, repo string, checkRunID int64) (*CheckRun, *Response, error) {
+	u := fmt.Sprintf("repos/%v/%v/check-runs/%v", owner, repo, checkRunID)
+	req, err := s.client.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	acceptHeaders := []string{mediaTypeCheckRunsPreview}
+	req.Header.Set("Accept", strings.Join(acceptHeaders, ", "))
+
+	checkRun := new(CheckRun)
+	resp, err := s.client.Do(ctx, req, checkRun)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return checkRun, resp, nil
+}
+
+// CreateCheckRunOptions sets up parameters need to create a CheckRun.
+type CreateCheckRunOptions struct {
+	Name        *string         `json:"name,omitempty"`
+	HeadBranch  *string         `json:"head_branch,omitempty"`
+	HeadSHA     *string         `json:"head_sha,omitempty"`
+	DetailsURL  *string         `json:"details_url,omitempty"`
+	ExternalID  *int64          `json:"external_id,omitempty"`
+	Status      *string         `json:"status,omitempty"`
+	Conclusion  *string         `json:"conclusion,omitempty"`
+	StartedAt   *Timestamp      `json:"started_at,omitempty"`
+	CompletedAt *Timestamp      `json:"completed_at,omitempty"`
+	Output      *CheckRunOutput `json:"output,omitempty"`
+}
```

Comment:
And this struct should be modified so the required parameters are values, not pointers. They also don't need `,omitempty` option. So:

```Go
// CreateCheckRunOptions sets up parameters need to create a CheckRun.
type CreateCheckRunOptions struct {
	Name        string  `json:"name"`                  // The name of the check (e.g., "code-coverage"). (Required.)
	HeadBranch  string  `json:"head_branch"`           // The name of the branch to perform a check against. (Required.)
	HeadSHA     string  `json:"head_sha"`              // The SHA of the commit. (Required.)
	DetailsURL  *string `json:"details_url,omitempty"` // The URL of the integrator's site that has the full details of the check. (Optional.)
	ExternalID  *int64  `json:"external_id,omitempty"` // A reference for the run on the integrator's system. (Optional.)
	...
}
```

Search the codebase for "(Required.)" or "(Optional.)" to see examples of this pattern. E.g.:

https://github.com/google/go-github/blob/8ea2e2657df890db8fb434a9274799d641bd698c/github/orgs_teams.go#L116-L122

Reply:
@shurcooL Thanks for the extra 👀 and pointers.  Any recommended pattern to follow when a `required` field is dependent on another field. For example the `conclusion` field has
> Required if you provide a status of completed.

```

conclusion | string | Required. The final conclusion of the check. Can be one of success, failure, neutral, cancelled, timed_out, or action_required. When the conclusion is action_required, additional details should be provided on the site specified by details_url. Required if you provide a status of completed.
```

The same for the `completed_at` field
>Required if you provide conclusion.

Ideally these would be an `omitempty` by default but shouldn't be when certain other fields are present.

Reply:
> Any recommended pattern to follow when a required field is dependent on another field.

Yes, we have no choice but to make them a pointer with `,omitempty`, and just rely on documentation.

See here for precedent:

https://github.com/google/go-github/blob/8ea2e2657df890db8fb434a9274799d641bd698c/github/projects.go#L67-L68

(It doesn't use a pointer... but that entire struct uses values for optional fields and expects them to be zero value when unset. Unfortunate that it's not consistent.)

Reply:
Thanks for that. I updated the struct to include required/optional as comments along side the field.